### PR TITLE
Fix some UB like memcpy(NULL, NULL, 0)

### LINKED ISSUE #25819: Some Undefined Behaviors like `memcpy(NULL, NULL, 0)`
#### Some reference about `memcpy(dest, src, count)`

According to https://en.cppreference.com/w/cpp/string/byte/memcpy, if either `dest` or `src` is an invalid or null pointer, the behavior is undefined, even if count is zero.

And there is an [example](https://godbolt.org/z/aKWhsMrbb) shows how this UB affect program's behavior after `gcc -O3` optimization. 

## Expected Behavior
Eliminating this kind of UB.

### xbmc-xrandr.c
After returning from `init_transform` at line 1240, `&crtcs[c].current_transform`'s field `nparams` is 0, `params` is NULL. Then `&crtcs[c].current_transform` is passed as 2nd parameter of `copy_transform` at line 1242.

https://github.com/xbmc/xbmc/blob/741f0604fddf1a33aca00c88f6d148f3723c62e9/xbmc-xrandr.c#L1238-L1242
https://github.com/xbmc/xbmc/blob/741f0604fddf1a33aca00c88f6d148f3723c62e9/xbmc-xrandr.c#L582-L591

Entering `copy_transform`, parameter `src` correspond to `&crtcs[c].current_transform`, so `src->params` is NULL, `src->nparams` is 0.

https://github.com/xbmc/xbmc/blob/741f0604fddf1a33aca00c88f6d148f3723c62e9/xbmc-xrandr.c#L603-L606

Then `memcpy(dest->params, NULL, 0)` at line 600 is undefined behavior.

https://github.com/xbmc/xbmc/blob/741f0604fddf1a33aca00c88f6d148f3723c62e9/xbmc-xrandr.c#L593-L601

### xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxCC.cpp
If `data.size()` is 0, `pPacket->pData` is initialized as NULL, then there is a UB at line 488.

https://github.com/xbmc/xbmc/blob/741f0604fddf1a33aca00c88f6d148f3723c62e9/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxCC.cpp#L486-L488

### lib/libUPnP/Neptune/Source/Core/NptUtils.cpp
At line 370, if `separator` is NULL, `separator_length` is set to 0, then there is a UB at line 380.

https://github.com/xbmc/xbmc/blob/741f0604fddf1a33aca00c88f6d148f3723c62e9/lib/libUPnP/Neptune/Source/Core/NptUtils.cpp#L370-L383

## Possible Fix

Check `count` before invoking `memcpy`.

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [x] Linux
 - [ ] macOS
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name:
 - Kodi version:

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by hribz**
## Description

There are some undefined behaviors like `memcpy(NULL, NULL, 0)` in current source code.

## Motivation and context

Fix #25819 

## How has this been tested?

Just some trivial change.

## What is the effect on users?

Eliminating undefined behaviors make program more stable.

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: lib/libUPnP/Neptune/Source/Core/NptUtils.cpp [hunk 1]
```diff
@@ -376,7 +376,7 @@ NPT_HexString(const unsigned char* data,
     while (data_size--) {
         NPT_ByteToHex(*src++, dst, uppercase);
         dst += 2;
-        if (data_size) {
+        if (data_size && separator_length) {
```
Comment (enen92) [inline lib/libUPnP/Neptune/Source/Core/NptUtils.cpp:379]: This is an external library, can you please add the .patch file as well?

**Comment by lrusak** [comment]
The problem with this change is that is it valid to proceed if the memcpy is protected?

Should special handling be taken in that case that the non protected path gets taken or should we just throw an exception or trigger an assert?

File: xbmc/XBDateTime.cpp [hunk 2]
```diff
@@ -44,6 +44,8 @@ CDateTimeSpan::CDateTimeSpan(const CDateTimeSpan& span)
 
 CDateTimeSpan::CDateTimeSpan(int day, int hour, int minute, int second)
 {
+  m_timeSpan.highDateTime = 0;
+  m_timeSpan.lowDateTime = 0;
```
Comment (neo1973) [inline xbmc/XBDateTime.cpp:48]: It's simpler to use a delegating constructor like this IMO:
```cpp
CDateTimeSpan::CDateTimeSpan(int day, int hour, int minute, int second) : CDateTimeSpan()
{
  SetDateTimeSpan(day, hour, minute, second);
}
```

File: xbmc-xrandr.c [hunk 3]
```diff
@@ -597,7 +597,8 @@ static void set_transform(
   dest->filter = strdup(filter);
   dest->nparams = nparams;
   dest->params = malloc(nparams * sizeof(XFixed));
-  memcpy(dest->params, params, nparams * sizeof(XFixed));
+  if (nparams)
+    memcpy(dest->params, params, nparams * sizeof(XFixed));
```
Comment (neo1973) [inline xbmc-xrandr.c:601]: Calling `malloc` with size 0 is implementation defined. Better change it to this for predictable behavior:
```cpp
static void set_transform(
    transform_t* dest, XTransform* transform, char* filter, XFixed* params, int nparams)
{
  dest->transform = *transform;
  dest->filter = strdup(filter);
  dest->nparams = nparams;
  if (nparams)
  {
    dest->params = malloc(nparams * sizeof(XFixed));
    memcpy(dest->params, params, nparams * sizeof(XFixed));
  }
  else
    dest->params = NULL;
}
```

**Review by neo1973 — CHANGES_REQUESTED** [review]

**Comment by hribz** [comment]
> The problem with this change is that is it valid to proceed if the memcpy is protected?
> 
> Should special handling be taken in that case that the non protected path gets taken or should we just throw an exception or trigger an assert?

The main issue with the relevant code lies in the assumption that when `count` is 0, `memcpy(dest, src, count)` will do nothing, hence `dest` or `src` can be NULL. However, even if `count` is 0, having `dest` or `src` as NULL constitutes undefined behavior (#25819 has more specific explaination).

Therefore, I think that protect the `memcpy` function can eliminate this undefined behavior while maintaining its original semantics.

File: xbmc/dbwrappers/mysqldataset.cpp [hunk 4]
```diff
@@ -1634,7 +1634,10 @@ std::string MysqlDatabase::mysql_vmprintf(const char* zFormat, va_list ap)
 
   mysqlStrAccumInit(&acc, zBase, sizeof(zBase), MYSQL_MAX_LENGTH);
   mysqlVXPrintf(&acc, 0, zFormat, ap);
-  return mysqlStrAccumFinish(&acc);
+  std::string strResult = std::string(mysqlStrAccumFinish(&acc));
```
Comment (neo1973) [inline xbmc/dbwrappers/mysqldataset.cpp:1637]:
```suggestion
  std::string strResult = mysqlStrAccumFinish(&acc);
```

**Comment by neo1973** [comment]
Please remove the unnecessary "[Merge branch 'xbmc:master' into master](https://github.com/xbmc/xbmc/pull/25820/commits/73abb79eaea52f759d42228d000db18a6e947411)" commit.

**Comment by neo1973** [comment]
Jenkins build this please.

**Review by neo1973 — COMMENTED** [review]
The end result looks now good :+1:

What's not so nice are the commits, they are not structured properly. Please clean them up with one meaningful change per commit.

**Comment by hribz** [comment]
> The end result looks now good 👍
> 
> What's not so nice are the commits, they are not structured properly. Please clean them up with one meaningful change per commit.

I'm not very familiar with git. Could you please help me check if the new commit complies with the standards? Thank you!

**Review by neo1973 — APPROVED** [review]
Thank you, looks good to me :slightly_smiling_face:

**Pull request merged**